man/systemd-sysext.xml

```diff
@@ -0,0 +1,235 @@
+    <filename>/opt/</filename> hierarchies become read-only too while they are activated.</para>
+
+    <para>System extensions are supposed to be purely additive, i.e. they are supposed to include only files
+    that do not exist in the underlying basic OS image. However, the underlying mechanism (overlayfs) also
+    allows removing files, but it is recommended not to make use of this.</para>
+
+    <para>System extension images may be provided in the following formats:</para>
+
+    <orderedlist>
+      <listitem><para>Plain directories or btrfs subvolumes containing the OS tree</para></listitem>
+      <listitem><para>Disk images with a GPT disk label, following the <ulink
+      url="https://systemd.io/DISCOVERABLE_PARTITIONS">Discoverable Partition Specification</ulink></para></listitem>
+      <listitem><para>Disk images lacking a partition table, with a naked Linux file system (e.g. squashfs or ext4)</para></listitem>
+    </orderedlist>
+
+    <para>These image formats are the same ones that
+    <citerefentry><refentrytitle>systemd-nspawn</refentrytitle><manvolnum>1</manvolnum></citerefentry>
+    supports via it's <option>--directory=</option>/<option>--image=</option> switches and those that the
+    service manager supports via <option>RootDirectory=</option>/<option>RootImage=</option>. Similar to
+    them they may optionally carry Verity authentication information.</para>
+
+    <para>System extensions are automatically looked for in the directories
+    <filename>/etc/extensions/</filename>, <filename>/run/extensions/</filename>,
+    <filename>/var/lib/extensions/</filename>, <filename>/usr/lib/extensions/</filename> and
+    <filename>/usr/local/lib/extensions/</filename>. The first two listed directories are not suitable for
+    carrying large binary images, however are still useful for carrying symlinks to them. The primary place
+    for installing system extensions is <filename>/var/lib/extensions/</filename>. Any directories found in
+    these search directories are considered directory based extension images, any files with the
+    <filename>.raw</filename> suffix are considered disk image based extension images.</para>
+
+    <para>During boot OS extension images are activated automatically, if the
+    <filename>systemd-sysext.service</filename> is enabled. Note that this service runs only after the
+    underlying file systems are mounted where system extensions are discovered, i.e. relatively late during
+    boot. This means they are not suitable for shipping resources that are processed by subsystems running in
+    earlier boot. Specifically, OS extension images are not suitable for shipping system services or
+    <citerefentry><refentrytitle>systemd-sysusers</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+    definitions. See <ulink url="https://systemd.io/PORTABLE_SERVICES">Portable Services</ulink> for a simple
```

Comment:
I assume this is before sysinit.target, but that probably needs to be said explicitely.

Reply:
dunno, I'd leave this a bit vague regarding the precise deps. I have the suspicion we have to tweak them a bit still, and I'd rather not document something that is quite likely going to change in the near future. I mean, the important stuff is that it's too late for system services, and that I mention, but the rest I'd really leave vague — at least for now. hope that makes sense?

Reply:
I kind of understand your point, but the main usage I can see is to be able to update binaries... which are probably used by other services... so it makes it quite useless if I am not sure when it's available. 

Well, I guess i'll wait until this gets settled then

Is this plugged into RequiresMountFor in any way ? that would be a handy/safe way to have most users ordered properly via default dependencies...

Reply:
So, I added a line now that clarifies that by the time regular services initialize (i.e. by `basic.target` is reached, i.e. whenDefaultDependencies=yes services initialize) the extensions are guaranteed to be active. This gives us enough freedom to still shift things around during early-boot, but should give people enough of a guarantee that "regular" stuff can just access the resources the extensions provide.

Reply:
> Is this plugged into RequiresMountFor in any way ? that would be a handy/safe way to have most users ordered properly via default dependencies...

So, I though about doing that, but I was afraid that people establish mounts inside of the extensions dir (to mount individual images from elsewhere), and to catch that `RequiresMountsFor=` can't really work generically, but ordering things after local-fs.target does work, hence I opted for the latter, just to be safe here.